src/Panel.js

```diff
@@ -1,236 +1,143 @@
 import classNames from 'classnames';
-import React, { cloneElement } from 'react';
 import PropTypes from 'prop-types';
+import React from 'react';
+import uncontrollable from 'uncontrollable';
 
-import Collapse from './Collapse';
-import { bsStyles, bsClass, getClassSet, prefix, splitBsPropsAndOmit }
+import { bsStyles, bsClass, getClassSet, splitBsPropsAndOmit }
   from './utils/bootstrapUtils';
 import { State, Style } from './utils/StyleConfig';
+import Body from './PanelBody';
+import Heading from './PanelHeading';
+import Title from './PanelTitle';
+import Footer from './PanelFooter';
+import Toggle from './PanelToggle';
+import Collapse from './PanelCollapse';
 
-// TODO: Use uncontrollable.
+const defaultGetId = (id, type) => id ? `${id}--${type}` : null;
 
 const propTypes = {
-  collapsible: PropTypes.bool,
-  onSelect: PropTypes.func,
-  header: PropTypes.node,
-  id: PropTypes.oneOfType([
-    PropTypes.string, PropTypes.number,
-  ]),
-  footer: PropTypes.node,
-  defaultExpanded: PropTypes.bool,
+  /**
+   * Controls the collapsed/expanded state ofthe Panel. Requires a `Panel.Collapse` or `<Panel.Body collapsible>` child component
+   * in order to actually animate out or in,
+   *
+   * @controllable onToggle
+   */
   expanded: PropTypes.bool,
+  /**
+   * A callback fired when the collapse state changes.
+   *
+   * @controllable expanded
+   */
+  onToggle: PropTypes.func,
   eventKey: PropTypes.any,
-  headerRole: PropTypes.string,
-  panelRole: PropTypes.string,
-
-  // From Collapse.
-  onEnter: PropTypes.func,
-  onEntering: PropTypes.func,
-  onEntered: PropTypes.func,
-  onExit: PropTypes.func,
-  onExiting: PropTypes.func,
-  onExited: PropTypes.func,
-};
 
-const defaultProps = {
-  defaultExpanded: false,
+  /**
+   * An HTML `id` attribute uniquely identifying the Panel component.
+   */
+  id: PropTypes.string,
 };
 
-class Panel extends React.Component {
-  constructor(props, context) {
-    super(props, context);
-
-    this.handleClickTitle = this.handleClickTitle.bind(this);
-
-    this.state = {
-      expanded: this.props.defaultExpanded,
-    };
-  }
+const contextTypes = {
+  $bs_panelGroup: PropTypes.shape({
+    getId: PropTypes.func,
+    activeKey: PropTypes.any,
```

Comment:
do you generically prefer this context-based pattern now for passing through active state? should we be using this for `<Nav>` instead of having `<Nav>` clone all its children with an `active` (and `activeKey` and `activeHref`) prop?